### ISSUE TITLE
GPU Instancing Support

### ISSUE THREAD

**Issue by wdc63** [helix-toolkit/helix-toolkit#2024]
I am aiming to develop non-gaming 3D visualization applications, an industrial simulation tools, leveraging WPF and other UI frameworks. Currently, the open-source software options are quite limited (excluding Unity due to substantial licensing restrictions), and HelixToolkit seems to be a very promising choice. However, it appears that HelixToolkit may not support some advanced visualization features at the moment. Advanced GPU Instancing features like Skeleton and Mesh Skinning are crucial for certain simulation tools, for instance, simulating Crowd Walking. I am wondering if HelixToolkit plans to support these GPU Instancing methods in future releases, or if it already has such support in the current version.

**Comment by holance** [comment]
https://github.com/helix-toolkit/helix-toolkit/wiki/WPF-vs-WPF.SharpDX

**Comment by wdc63** [comment]
> https://github.com/helix-toolkit/helix-toolkit/wiki/WPF-vs-WPF.SharpDX
> 
> If you want to have different instances doing their own mesh skinning, then it's not being supported at this moment.

Yes, instancing bone skins or skeleton morphs something like that inside the mesh, is that in plans?

**Comment by holance** [comment]
How many instances are you trying to render?

**Comment by wdc63** [comment]
> How many instances are you trying to render?

May be around 10000, animate human walking, several types of human with different gender, age.. something like that.

**Comment by holance** [comment]
For this kind of simulation, you probably want to use commercial engine since the simulation data generation are engine based and need different kind of tools to create them

**Comment by wdc63** [comment]
> For this kind of simulation, you probably want to use commercial engine since the simulation data generations are engine based and need different kind of tools to create them. There are free engines like godot you can take a look.

Thx, but We have already developed a pedestrian library ourselves and have used helixtoolkit to achieve 2D visualization. Currently, we hope to continue using helixtoolkit to realize the 3D visualization of pedestrian movement and indoor scenes.
![Screenshot_2023-10-07-01-32-10-79_e39d2c7de19156b0683cd93e8735f348](https://github.com/helix-toolkit/helix-toolkit/assets/14074644/e8700c41-8335-45f9-bb1f-d764870c45e9)
![Screenshot_2023-10-07-01-31-21-63_e39d2c7de19156b0683cd93e8735f348](https://github.com/helix-toolkit/helix-toolkit/assets/14074644/8c99d497-abf3-4900-9580-855cf8bbab38)

**Comment by holance** [comment]
Do you need different animations for each human instance? Or can use same animation for all human instances?

**Comment by wdc63** [comment]
> Do you need different animations for each human instance? Or can use same animation for all human instances?

Pedestrians primarily have two main states: walking and standing. It would be great if the sitting could also be added. However, they have different speeds, so each pedestrian instance may exhibit slightly different movements. For example, some may have their left foot forward while others have their right foot forward at a given moment. Each pedestrian is in a different frame state within the animation. Therefore, I believe the answer should be that each pedestrian instance uses a different animation.

**Comment by holance** [comment]
Usually for this kind of animation, animation data need to be baked into texture and having an animation system to track animation state for each instance. And I don't think Helix toolkit will ever support it with limited resources.

**Comment by wdc63** [comment]
> Usually for this kind of animation, animation data need to be baked into texture and having an animation system to track animation state for each instance. And I don't think Helix toolkit will ever support it with limited resources.

There are many libraries for GPU instancing on GitHub, including several implementations that utilize SharpDX. I believe this is a very common technique. You can find some examples here:

Render-Crowd-Of-Animated-Characters: https://github.com/chenjd/Render-Crowd-Of-Animated-Characters
GPUInstance: https://github.com/mkrebser/GPUInstance
MonoGame-HardwareInstancing: https://github.com/PodeCaradox/MonoGame-HardwareInstancing
Skirmish: https://github.com/Selinux24/Skirmish

**Comment by holance** [comment]
The first two links are using unity. The last two links are doing non animated instancing which already supports in Helix toolkit.

**Comment by wdc63** [comment]
> The first two links are using unity. The last two links are doing non animated instancing which already supports in Helix toolkit.

Yes, currently Skinned instancing techniques are primarily focused on game engines. In the last mentioned repository, Skirmish, they have implemented Skinned instancing using SharpDX. They have implemented dynamic instancing for an army. You can find the implementation here: 
https://github.com/Selinux24/Skirmish/tree/dev/Samples/02%20IntermediateSamples/SceneInstancing.
